sklearn/model_selection/_validation.py

```diff
@@ -625,8 +625,12 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None,
                       pre_dispatch='2*n_jobs', method='predict'):
     """Generate cross-validated estimates for each input data point
 
-    It is not appropriate to pass these predictions into an evaluation
-    metric. Use :func:`cross_validate` to measure generalization error.
+    Using these predictions to measure the generalization performance of the model 
```

Comment:
I think this is not quite right: evaluating performance with these predictions is not, in general, correctly measuring generalisation performance. It is only under specific conditions.

Perhaps, then "Passing these predictions into an evaluation metric may not be a valid way to measure generalization performance. Results can differ from `cross_validate` unless all tests sets have equal size and the metric decomposes over samples."

Reply:
I think that using cross_validate is also not, in general, a good way of measuring generalization performance: it requires specific characteristics of the splitting. It highly depends on the size of the testing sets and on the similarity between those sets (between them and with regard to the training set).

Reply:
What do you mean? `cross_validate` is exactly the same as `cross_val_score` apart from that you can give several metrics and you get timing information. The splitting is done through the `cv` objects similarly

Reply:
I just meant that using `cross_val_predict` predictions to compute generalization performance has its own drawbacks but that the averaging estimation of performances done in` cross_val_score` and `cross_validate` also makes strong assumption on the testing sets.

Reply:
Could you clarify what you mean by strong assumptions about test sets?  In both cases, I think, the user needs to select cv carefully.

Reply:
Yes I totally agree: both cases require to be careful about the conclusions made on generalization. In the averaging, we assume equally sized test sets with similar distribution of classes

Reply:
>  In the averaging, we assume equally sized test sets with similar distribution of classes

I disagree. In averaging we assume that the sample is similar to random. Small test sets are tricky, but as a rule, if there is high variance among the per-test-set scores, it deserves investigation.

Reply:
I agree that high variance deserves investigation. But don't we assume during cross validation that each fold should have the same size ?

Reply:
>  But don't we assume during cross validation that each fold should have the same size ?

Well from an implementation side, we certainly don't require it in scikit-learn, and it's far from assured in something like LeaveOneGroupOut. I don't really see how it provides much benefit to require them to have the same size, as long as all are of sufficient size for the score measurement to be reasonably accurate. But I'm not a statistician, and may not appreciate the underlying theory.

Reply:
I am not either, but to me, estimation by averaging of folds of different sizes seems to be a biased estimation of performances. I am ok to just put a warning to let the user decide how to handle the problem but this warning should also be added to the `cross_validation` and `cross_val_score` functions